velox/connectors/hive/iceberg/IcebergSplitReader.cpp

```diff
@@ -47,8 +51,22 @@ IcebergSplitReader::IcebergSplitReader(
           fileHandleFactory,
           executor,
           scanSpec),
+      originalScanSpec_(nullptr),
+      originalRemainingFilters_(nullptr),
+      remainingFilterExprSet_(remainingFilterExprSet),
       baseReadOffset_(0),
-      splitOffset_(0) {}
+      splitOffset_(0),
+      expressionEvaluator_(expressionEvaluator) {}
+
+IcebergSplitReader::~IcebergSplitReader() {
+  if (originalScanSpec_) {
+    *scanSpec_ = *originalScanSpec_;
```

Comment:
This is suboptimal as we lose the filter ordering adaptivity.  The filter ordering adaptivity is only in `scanSpec_` and dynamic filter push down is only in `originalScanSpec_`.  We can solve this by instead of copying the whole spec tree, we push in additional filter onto the specific node, and pop it when we are done with the delta split:

```c++
class ScanSpec {
  Filter* filter() const {
    return filters_.empty() ? nullptr : filters_.back().get();
  }  

  void pushFilter(std::unique_ptr<Filter> newFilter) {
    if (!filter_.empty()) {
      newFilter->mergeWith(filter_.back().get());
    }
    filters_.push_back(std::move(newFilter));
  }

  void popFilter() {
    filters_.pop_back();
  }

  // A stack of filters, the upper ones are more restrictive than the lower ones in the stack, and we only use the topmost one during reading. When dynamic pushdown comes in, we apply to all filters in the stack.  When a equality delta delete split is finished, we pop the top filter representing the delta delete.
  std::vector<std::unique_ptr<Filter>> filters_;
};
```

Reply:
> This is suboptimal as we lose the filter ordering adaptivity. The filter ordering adaptivity is only in scanSpec_ and dynamic filter push down is only in originalScanSpec_. We can solve this by instead of copying the whole spec tree, we push in additional filter onto the specific node, and pop it when we are done with the delta split:

@Yuhta Thanks for pointing out that the filter ordering adaptivity is lost. Your idea is very smart! So the `filters_` vector holds all versions of the filter, and we always use the latest version. Just a minor nit, I am thinking to rename `pushFilter` to `updateFilter` and `popFilter` to `restoreFilter` so that they are easier to understand.

What you haven't mentioned was the fields themselves may need to be inserted and deleted too, since the new filters may contain fields that are not already in the ScanSpec. For that we need to support removing/restoring ScanSpec nodes.  We can either add a boolean deleteFlag in each ScanSpec node, or just directly remove it from the children_ vector directly. If you think this is good I'll add this as well.